keps/sig-instrumentation/draft-20190404-kubernetes-control-plane-metrics-stability.md

```diff
@@ -0,0 +1,252 @@
+    Name: "some_alpha_metric",
+    Help: "some description",
+    StabilityLevel: kubemetrics.ALPHA, // this is also a custom metadata field
+}
+```
+
+### Metric Instantiation Phase
+
+Currently, the metric instantiation phase looks like this:
+
+```go
+var someCounterVecMetric = prometheus.NewCounterVec(
+    someMetricDefinition,
+    []string{"some-label", "other-label"},
+}
+```
+
+Wrapping the prometheus constructors would allow us to take, as inputs, the modified metric definitions defined above, returning a custom kubernetes metric object which contains the metric which would have been instantiated as well as the custom metadata:
+
+```go
+var deprecatedMetric = kubemetrics.NewCounterVec( // this is a wrapped initializer, which takes in our custom metric definitions
+    deprecatedMetricDefinition, // this is our custom wrapped metric definition from above
+    []string{"some-label", "other-label"},
+}
+var alphaMetric = kubemetrics.NewCounterVec{
+    alphaMetricDefinition, // this is also our custom wrapped metric definition from above
+    []string{"some-label", "other-label"},
+}
+```
+
+### Metric Enrollment Phase
+
+Currently, metric enrollment involves calls to a prometheus function which enrolls the metric in a global registry, like so:
+```go
+prometheus.MustRegister(someCounterVecMetric)
+```
+
+Wrapping a prometheus registry with a kubernetes specific one, would allow us to take our custom metrics from our instantiation phase and execute custom logic based on our custom metadata. Our custom registry would hold a reference to a prometheus registry and defer metric enrollment unless preconditions were met:
+
+```go
+import version "k8s.io/apimachinery/pkg/version"
+
+type Registry struct {
+    promregistry *prometheus.Registry
+    KubeVersion version.Info
+}
+
+// inject custom registration behavior into our registry wrapper
+func (r *Registry) MustRegister(metric kubemetrics.Metric) {
+    // pretend we have a version comparison utility library
+    if metricutils.compare(metric.DeprecatedVersion).isLessThan(r.KubeVersion) {
+        // check if binary has deprecated metrics enabled otherwise
+        // no-op registration
+        return
+    } else if metricutils.compare(metric.DeprecatedVersion).isEqual(r.KubeVersion) {
+        // append deprecated text to description
+        // emit warning in logs
+        // continue to actual registration
+    }
+    // append alpha text to metric description if metric.isAlpha
+    // fallback to original prometheus behavior
+    r.promregistry.MustRegister(metric.realMetric)
+}
+
+```
+
+Which we would invoke, like so:
+```go
+kubemetrics.MustRegister(deprecatedMetric)
+kubemetrics.MustRegister(alphaMetric)
+```
+
+## Stability Classes
+
+This proposal introduces three stability classes for metrics: (1) Alpha, (2) Stable, (3) Deprecated. These classes are intended to make explicit the API contract between the control-plane and the ingester of control-plane metrics.
```

Comment:
Should we consider Deprecated orthogonal to the stability class?  I.E. metrics can be Alpha or Stable, and Deprecated or Non-Deprecated?

While we are always free to remove alpha metrics, there are often cases where I as a component owner want to make the transition easier for users, and want to have a deprecation period even though I am not contractually required to.

Reply:
I really like the idea!